### ISSUE TITLE
[Question] How can i ignore initialRoute when using with initialDeepLink?

### ISSUE THREAD

**Issue by cogivn** [Milad-Akarie/auto_route_library#1518]
Hi,

When I launched the application with `initialDeepLink`, the app displayed both pages that I configured with paths '/' and the path settled into `initialDeepLink`. How can I display only the page settled in `initialDeepLink` and ignore the page with the root path '/'? 

Thanks.

```dart
@singleton
@AutoRouterConfig(replaceInRouteName: 'Page,Route')
class AppRouter extends _$AppRouter {
  @override
  List<AutoRoute> get routes => [
        AutoRoute(page: WelcomeRoute.page, path: '/'),
        AutoRoute(page: TestRoute.page),
        AutoRoute(page: HomeRoute.page),
        AutoRoute(page: PostPreviewRoute.page),
        AutoRoute(page: PerkDetailRoute.page),
        AutoRoute(page: EventReservationRoute.page, path:'/event-reservation/:id'),
        AutoRoute(page: EventDetailRoute.page),
        AutoRoute(page: SearchRoute.page),
      ];
}
```

```dart
class AppWidget extends StatelessWidget {
  const AppWidget({super.key});

  @override
  Widget build(BuildContext context) {
    final router = getIt<AppRouter>();
    return BlocBuilder<LangCubit, Locale>(
      builder: (context, locale) {
        return AppThemeWrapper(
          appTheme: AppTheme.create(locale),
          builder: (context, themeData) => MaterialApp.router(
            localizationsDelegates: const [
              S.delegate,
              GlobalMaterialLocalizations.delegate,
              GlobalWidgetsLocalizations.delegate,
              GlobalCupertinoLocalizations.delegate,
            ],
            debugShowCheckedModeBanner: false,
            supportedLocales: S.delegate.supportedLocales,
            locale: locale,
            builder: (context, child) {
              final MediaQueryData data = MediaQuery.of(context);
              return MediaQuery(
                data: data.copyWith(textScaleFactor: 1.0),
                child: Asuka.builder(context, child),
              );
            },
            theme: themeData,
            routerConfig: router.config(initialDeepLink: '/event-reservation/20'),
          ),
        );
      },
    );
  }
}

```

**Comment by hasanmhallak** [comment]
hey @cogivn, you can use a `RouteGuard` to check your if your app started normally or from a deeplink like this:
```dart
  @override
  void onNavigation(NavigationResolver resolver, StackRouter router) {
    /// Wheather the app has been started from launcher
    /// or from a deeplink.
    final normalStart = resolver.pendingRoutes.isEmpty;
    
   // then you can push the right route accordingly.
```

You may find this tedious but in fact this is important as you may want for your user to continue with the normal flow of your application according to [Principles of navigation](https://developer.android.com/guide/navigation/navigation-principles#deep-link)

**Comment by cogivn** [comment]
Hi @hasanmhallak,

Thank you for sharing new useful information with me. 
I have some question about: 

1. if I want to start a new screen different from the screen I registered root when opening the application from the launcher(change initialRoute programmatically), how can I do that?
2. Is there any function that can help me find PageRouteInfo by path? For example: When I received the path '/event-reservation/20' and I want to look it up in the stack and return the EventReservationRoute().

Once again, thank you very much for your feedback :)

**Comment by Milad-Akarie** [comment]
@cogivn for your initial question you can ignore prefixed matches by setting 
`_appRouter.config(includePrefixMatches: false)` 
this will only add the fully matched path to the stack.
alternatively you can set fullMatch to true on the initial route

**Comment by Milad-Akarie** [comment]
@hasanmhallak can you talk more about this use-case, maybe I can come up with a better implementation for it?

**Comment by Milad-Akarie** [comment]
@cogivn is there a reason why you need to convert path PageRouteInfo your self? why not let the router handle that by using pushNamed or navigateNamed?

**Comment by cogivn** [comment]
Hi @Milad-Akarie,

I am developing a sub-app and trying to configure the root page based on the path passed in. 
I see that **auto_route** has 2 configurations that allow me to pass the initial path through `initialDeepLink` and `initialRoutes`. 

If I want to use `initialRoutes` through the path passed in from outside, then I need to find `PageRouteInfo` in the stack based on the path. That's why I'm asking if there is a way to get PageRouteInfo based on the path or not. 

If I use `pushNamed` or `navigateNamed` next to the root, it opens both the root page and the destination page at the same time, but I only want to open one of the pages depends on path.

Is there any other way to handle the above issue without using `inititalDeepLink` or `inititialRoutes`?

**Comment by Milad-Akarie** [comment]
@cogivn by sub-app do you mean it will never get a deep-link from platform?

**Comment by cogivn** [comment]
@Milad-Akarie Yes. 
The main app will handle the deep link and pass the path to the sub-app, and in the sub-app, it will navigate depending on the path passed in.

**Comment by Milad-Akarie** [comment]
@cogivn alright then you can manually set the initial-deep-link your self with no issues. 
as I said setting includePrfixMatches to false will only take you to target route without including the home route.

do you have any other issue?

**Comment by hasanmhallak** [comment]
@Milad-Akarie  the [Principles of navigation](https://developer.android.com/guide/navigation/navigation-principles#deep-link) by android say that if the user open a deep link he should be able to navigate back using the back button (up button if there is an appBar ) like if the he start the app the normal way and navigate to that route e.g:

let's say that our navigation graph look like this
```
- mainScreen
    - profileScreen
      - settingScreen
```

to get to the `settingScreen` in a normal flow you have to open the app which will load the `mainScreen` as initial route, then navigate to `profileScreen`, and then to `settingScreen`.

Now the principle says that if we need to create a deep link to the settingScreen. the RouteStack should be the same as the normal flow, meaning we should  push the whole RouteStack that can lead to the specific route the deep link will get us to. This way if the user uses that deep link and press the back button it will get back to the `profileScreen` and not exit the app.

**Comment by hasanmhallak** [comment]
@cogivn 
> if I want to start a new screen different from the screen I registered root when opening the application from the launcher(change initialRoute programmatically), how can I do that?

you can just push the route you want in the guard. though I'm not sure if you have to reject the initial call to guard first like:
```dart
 resolver.next(false);
router.push(YourRoute);
```
because I noticed that there are some Futures that has to completed. maybe @Milad-Akarie can clarify this point further more.

> Is there any function that can help me find PageRouteInfo by path? For example: When I received the path '/event-reservation/20' and I want to look it up in the stack and return the EventReservationRoute().

you can use `router.matcher.match(rawPath)` that will return a list of a `RouteMatch`, from there the `RouteMatch` has property to convert it to a `PageRouteInfo`:
```dart
RouteMatch().toPageRouteInfo();
```

**Comment by cogivn** [comment]
Thanks, @hasanmhallak. Let me try with your suggestion :)

**Comment by Milad-Akarie** [comment]
@hasanmhallak in case of rejecting the navigation you should always abort with next(false) otherwise you're leaving an completed completer waiting to be completed.

I totally agree about having the deep-link building a stack similar to what a manual navigation builds. auto_route was built with this in mind.
fyi auto_route can handle stacked-path matching as well e.g  if we have routes on the same level (not-nested) like follows:

```yaml
/
/products
/products/:id 
```
and we launch with a deep-link looking like this  `/products/123`

all above routes will be added to stack. 


@cogivn  @hasanmhallak  you can also do this 
`router.buildPageRoute(rawPath)`

**Comment by cogivn** [comment]
Hi @Milad-Akarie,

i see the library has been upgraded and I have configured in my case:
```dart
routerConfig: router.config(deepLinkBuilder: (deeplink) {
              return initialDeepLink != '/'
                  ? DeepLink.path(initialDeepLink)
                  : deeplink;
            }),
```

but can't go to previous page. I set a breakpoint for the debugger and then found that `router.pop()` always returns false because `_parent=null` and `disposition=bubble`. is this normal for `initialDeepLink` case?

And if this is the case how can I go back to the previous page in this case?

Thank you :D,
<img width="2061" alt="Screenshot 2023-04-28 at 07 40 34" src="https://user-images.githubusercontent.com/21217547/235028678-179a839f-c44e-4f9c-9f4e-7c300500410e.png">
<img width="2060" alt="Screenshot 2023-04-28 at 07 41 39" src="https://user-images.githubusercontent.com/21217547/235028698-77d4bfc5-cd92-4295-ba43-44acbe29a026.png">
